driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -1170,55 +1183,270 @@ int getNumPinnedToTransaction() {
         }
     }
 
+    /**
+     * This class maintains threads needed to perform {@link #getAsync(SingleResultCallback)}.
+     * This operation includes creating an {@link InternalConnection} ({@code create})
+     * and {@linkplain InternalConnection#open() opening} ({@code open}) it.
+     * These tasks may be blocked due to the {@link ConnectionPoolSettings#getMaxSize() max size} and
+     * {@link OpenConcurrencyLimiter#OpenConcurrencyLimiter(int) max connecting} limits respectively.
+     * An {@code open} task blocked because of the max connecting limit is also being counted towards the max size limit,
+     * i.e., it may block a {@code create} task. If we were to execute both kinds of tasks in the same thread, then it would
+     * totally order all tasks, making it possible for a {@code create} task to block an {@code open} task and, thus, open opportunities for
+     * tasks deadlocking. As a result, we must maintain at least two sets of threads for two kinds of tasks.
+     * <p>
+     * Due to tasks having a timeout, we also must monitor queued tasks to determine which ones timed out and completing them accordingly.
+     * While this logic can be incorporated in {@code create}/{@code open} tasks, doing so would have significantly complicate them further,
+     * and it is especially not obvious how to maintain fairness in that case.
+     */
     @ThreadSafe
-    private static class Workers implements AutoCloseable {
-        private volatile ExecutorService getter;
-        private volatile ExecutorService opener;
+    private static class AsyncWorkManager implements AutoCloseable {
```

Comment:
Given all the new synchronization logic in this class, should we add a stress/concurrency unit test to detect subtle bugs or race conditions?

Reply:
While thinking about execution in which the timeout handling was broken, I realized that things can (and must) be much better that they currently are. I changed this PR to the draft status and opened https://github.com/mongodb/mongo-java-driver/pull/722. The commit messages there explain everything, I am also happy to discuss the matter over a call if that would be simpler.